src/v/utils/tests/coro_bench.cc

```diff
@@ -0,0 +1,245 @@
+#include "base/seastarx.h"
+#include "ssx/future-util.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/testing/perf_tests.hh>
+#include <seastar/util/later.hh>
+
+#include <string>
+
+bool always_false_coro = false;
+
+namespace {
+
+static constexpr size_t ITERS = 10000;
+
+template<typename F>
+ss::future<size_t> co_await_in_loop(F f) {
+    perf_tests::start_measuring_time();
+    for (int i = 0; i < ITERS; i++) {
+        co_await f();
+    }
+    perf_tests::stop_measuring_time();
+    co_return ITERS;
+}
+
+template<typename F>
+ss::future<size_t> collect_futures(F f) {
+    std::vector<ss::future<>> futs;
+    futs.reserve(ITERS);
+    perf_tests::start_measuring_time();
+    for (int i = 0; i < ITERS; i++) {
+        futs.emplace_back(f());
+    }
+    perf_tests::stop_measuring_time();
+
+    for (auto& fut : futs) {
+        co_await std::move(fut);
+    }
+    co_return ITERS;
+}
+
+[[gnu::noinline]] ss::future<> ss_now() { return ss::now(); }
+
+[[gnu::noinline]] ss::future<> empty_cont() { return ss::maybe_yield(); }
+
+[[gnu::noinline]] static auto ready_future() {
+    return ss::make_ready_future<int>(1);
+}
+
+ss::future<int> empty_coro() {
+    int x = 0;
+    if (always_false_coro) {
```

Comment:
Won't this get optimized away? Or is this the point?

Reply:
It does not get optimized away, at least with our current compiler settings. The point is to simulate a coroutine which always returns ready, but where the compiler can't prove that is the case, which I think is the most interesting case:

1) The case where the compiler can prove that the future/awaitable returned never actually causes suspension can cause the whole thing to be complied/inlined as a plain function without any of the coroutine costs at all. This would seem to be very rare and probably vanishingly rare in the case of coroutines (because of the implicit preemption check) since those things don't need to be async at all.

2) The case where the coroutine actually suspends is common enough dominated by that machinery (scheduling a task, running said-task) rather than the costs we are probing here: that looks like roughly 1000 cycles more or less.

This `always_false_coro` is actually at a bit of risk of being compiled out if LTO is enabled and the compiler can see this value is never modified in another TU as so is effectively constant. I should actually make this a function and use `perf_tests::do_not_optimize` which is resistant to LTO (and also leads to less overhead in the actual check).